swarm/src/protocols_handler/one_shot.rs

```diff
@@ -72,7 +72,7 @@ where
             dial_queue: SmallVec::new(),
             dial_negotiated: 0,
             max_dial_negotiated: 8,
-            keep_alive: KeepAlive::Yes,
+            keep_alive: KeepAlive::Until(Instant::now() + config.keep_alive_timeout),
```

Comment:
It may be better to just move this condition from `inject_fully_negotiated_outbound`
```rust
        if self.dial_negotiated == 0 && self.dial_queue.is_empty() {
            self.keep_alive = KeepAlive::Until(Instant::now() + self.config.keep_alive_timeout);
        }
```
into `poll()`
```rust
       ...
       if !self.dial_queue.is_empty() {
            ...
        } else {
            self.dial_queue.shrink_to_fit();
            if self.keep_alive.is_yes() && self.dial_negotiated == 0 {
                self.keep_alive = KeepAlive::Until(Instant::now() + self.config.keep_alive_timeout);
            }
        }
```
The reason being that starting timers on creation of connection handlers is mildly problematic, because at least right now, handlers are created before the connection is even established, whereas a connection handler is `poll()`ed as soon as it is established and it gives the handler a chance to "make its first step" before the timeout starts (if there is indeed nothing that the handler wishes to do, as determined by `poll()`).